File: Software/NodeJS/libs/sensors/base/analogSensor.js
var util         = require('util')
var Sensor       = require('./sensor')
var commands     = require('../../commands')

function AnalogSensor(pin) {
  Sensor.apply(this, Array.prototype.slice.call(arguments))
  this.pin = pin
}
util.inherits(AnalogSensor, Sensor)
AnalogSensor.prototype = new AnalogSensor()

AnalogSensor.prototype.read = function(length) {
  if (typeof length == 'undefined')
    length = this.board.BYTESLEN

  var writeRet = this.board.writeBytes(commands.aRead.concat([this.pin, commands.unused, commands.unused]))
  if (writeRet) {
    this.board.readByte()
    var bytes = this.board.readBytes(length)
    if (bytes instanceof Buffer) {
      return bytes[1] * 256 + bytes[2]
    } else {
      return false
    }
  } else {
    return false
  }
}
AnalogSensor.prototype.write = function(value) {
  return this.board.writeBytes(commands.aWrite.concat([this.pin, value, commands.unused]))
}

module.exports = AnalogSensor

File: Software/Java/src/com/dexterind/grovepi/sensors/base/AnalogSensor.java
package com.dexterind.grovepi.sensors.base;

import java.io.IOException;

import com.dexterind.grovepi.*;
import com.dexterind.grovepi.utils.*;

public class AnalogSensor extends Sensor {
  protected int pin = 0;
  private int length = 4;

  protected Debug debug;

  public AnalogSensor(int pin, int length) throws IOException, InterruptedException, Exception {
	super();
	this.pin = pin;
	this.length = length == -1 ? this.length : length;
  }
  
  public byte[] readBytes() throws IOException {
	this.board.writeI2c(Commands.AREAD, this.pin, Commands.UNUSED, Commands.UNUSED);
	this.board.readI2c(1);
	return this.board.readI2c(this.length);
  }
  
  public boolean write(int value) throws IOException {
	this.board.writeI2c(Commands.AWRITE, this.pin, value, Commands.UNUSED);
	return true;
  }
}
